source/common/router/header_parser.cc

```diff
@@ -32,11 +35,41 @@ enum class ParserState {
 
 std::string unescape(absl::string_view sv) { return absl::StrReplaceAll(sv, {{"%%", "%"}}); }
 
+HttpHeaderFormatterPtr
+parseHttpHeaderFormatter(const envoy::config::core::v3::HeaderValue& header_value) {
+  const std::string& key = header_value.key();
+  // PGV constraints provide this guarantee.
+  ASSERT(!key.empty());
+  // We reject :path/:authority rewriting, there is already a well defined mechanism to
+  // perform this in the RouteAction, and doing this via request_headers_to_add
+  // will cause us to have to worry about interaction with other aspects of the
+  // RouteAction, e.g. prefix rewriting. We also reject other :-prefixed
+  // headers, since it seems dangerous and there doesn't appear a use case.
+  // Host is disallowed as it created confusing and inconsistent behaviors for
+  // HTTP/1 and HTTP/2. It could arguably be allowed on the response path.
+  if (!Http::HeaderUtility::isModifiableHeader(key)) {
+    throw EnvoyException(":-prefixed or host headers may not be modified");
+  }
+
+  // UPSTREAM_METADATA and DYNAMIC_METADATA must be translated from JSON ["a", "b"] format to colon
+  // format (a:b)
+  std::string final_header_value = HeaderParser::translateMetadataFormat(header_value.value());
+  // Change PER_REQUEST_STATE to FILTER_STATE.
+  final_header_value = HeaderParser::translatePerRequestState(final_header_value);
```

Comment:
Do we output warnings for deprecated usages? This is going to break someone in the future if/when any of the deprecated operators get used?

Reply:
I was planning to add warnings when runtime guard is retired. There are translators from "old" format and they will be kept for about 1 year from the moment warnings are added. After that, the "old" format will be completely deprecated. Is this OK?

Reply:
I would just add them now since the runtime guard defaults to enabled.

Reply:
Added. Thanks!